### ISSUE TITLE
Websocket blockingWrite hangs forever

### ISSUE THREAD

**Issue by superbazza** [eclipse/jetty.project#4781]
**Jetty version** : 9.4.10.v20180503

**Java version** : jdk1.8.0_231

**OS type/version** : Linux RHEL 2.6.32-754.27.1.el6.x86_64

**Description**
Process that sends websocket updates ended up hung for 24 hours. Aside from the logs, I don't have a lot to go on.

The following is the stack (from thread-dump) of the deadlocked thread:

```
    java.lang.Thread.State: WAITING (parking)
       at sun.misc.Unsafe.park(Native Method)
       - parking to wait for  <0x00000000b2d3c400> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
       at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
       at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
       at org.eclipse.jetty.util.SharedBlockingCallback$Blocker.block(SharedBlockingCallback.java:210)
       at org.eclipse.jetty.websocket.common.BlockingWriteCallback$WriteBlocker.block(BlockingWriteCallback.java:90)
       at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.blockingWrite(WebSocketRemoteEndpoint.java:107)
       at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.sendString(WebSocketRemoteEndpoint.java:394)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdatesWebSocket.sendMessage(DashboardUpdatesWebSocket.java:60)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor.lambda$sendTotal$2(DashboardUpdateProcessor.java:61)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor$$Lambda$199/1429247872.accept(Unknown Source)
       at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
       at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1082)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor.sendTotal(DashboardUpdateProcessor.java:61)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor.access$200(DashboardUpdateProcessor.java:11)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor$1.onDashboardTotalChange(DashboardUpdateProcessor.java:35)
       at com.myapp.storage.action.dashboard.DashboardTotal.sendUpdate(DashboardTotal.java:44)
       at com.myapp.storage.action.dashboard.Dashboard.sendUpdates(Dashboard.java:61)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor.processDashboard(DashboardUpdateProcessor.java:20)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateManager.lambda$checkForCellUpdates$0(DashboardUpdateManager.java:104)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateManager$$Lambda$186/849411958.accept(Unknown Source)
       at java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1597)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateManager.checkForCellUpdates(DashboardUpdateManager.java:94)
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateManager$UpdateDashboardsJob.execute(DashboardUpdateManager.java:154)
       at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
       at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
       - locked <0x00000000acde6970> (a java.lang.Object)

```

There were 2 errors preceding this deadlock situation:
```
 2020-04-15 11:02:56,448 [qtp1771407294-4763] WARN  c.b.f.s.h.s.d.u.DashboardUpdatesWebSocket - Unhandled Error (closing connection)
 java.lang.IllegalStateException: Blocking message pending 10000 for BLOCKING
       at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.lockMsg(WebSocketRemoteEndpoint.java:130) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.sendString(WebSocketRemoteEndpoint.java:386) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdatesWebSocket.sendMessage(DashboardUpdatesWebSocket.java:60) ~[prospero-2.37.0.jar:2.37.0]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdatesWebSocket.onText(DashboardUpdatesWebSocket.java:54) ~[prospero-2.37.0.jar:2.37.0]
       at sun.reflect.GeneratedMethodAccessor55.invoke(Unknown Source) ~[na:na]
       at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_231]
       at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_231]
       at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:71) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:72) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver.onTextMessage(JettyAnnotatedEventDriver.java:234) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.message.SimpleTextMessage.messageComplete(SimpleTextMessage.java:69) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.appendMessage(AbstractEventDriver.java:66) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver.onTextFrame(JettyAnnotatedEventDriver.java:226) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.incomingFrame(AbstractEventDriver.java:162) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.WebSocketSession.incomingFrame(WebSocketSession.java:476) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.AbstractExtension.nextIncomingFrame(AbstractExtension.java:183) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.compress.PerMessageDeflateExtension.nextIncomingFrame(PerMessageDeflateExtension.java:105) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.compress.CompressExtension.forwardIncoming(CompressExtension.java:142) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.compress.PerMessageDeflateExtension.incomingFrame(PerMessageDeflateExtension.java:85) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.incomingFrame(ExtensionStack.java:220) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.Parser.notifyFrame(Parser.java:220) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.Parser.parse(Parser.java:245) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.readParse(AbstractWebSocketConnection.java:560) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:391) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281) [jetty-io-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102) [jetty-io-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118) [jetty-io-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:760) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:678) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at java.lang.Thread.run(Thread.java:748) [na:1.8.0_231]
 2020-04-15 11:02:56,452 [MEERKAT-SERVER-QUARTZ-SERVER-CORE-LON_Worker-9] WARN  o.e.j.w.c.e.c.CompressExtension -
 java.lang.NullPointerException: Deflater has been closed
       at java.util.zip.Deflater.ensureOpen(Deflater.java:559) ~[na:1.8.0_231]
       at java.util.zip.Deflater.deflate(Deflater.java:440) ~[na:1.8.0_231]
       at org.eclipse.jetty.websocket.common.extensions.compress.CompressExtension$Flusher.compress(CompressExtension.java:488) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.compress.CompressExtension$Flusher.deflate(CompressExtension.java:451) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.compress.CompressExtension$Flusher.process(CompressExtension.java:431) ~[websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.compress.CompressExtension.outgoingFrame(CompressExtension.java:218) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.ExtensionStack$Flusher.process(ExtensionStack.java:419) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) [jetty-util-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.outgoingFrame(ExtensionStack.java:296) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.uncheckedSendFrame(WebSocketRemoteEndpoint.java:311) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.blockingWrite(WebSocketRemoteEndpoint.java:106) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.sendString(WebSocketRemoteEndpoint.java:394) [websocket-common-9.4.10.v20180503.jar:9.4.10.v20180503]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdatesWebSocket.sendMessage(DashboardUpdatesWebSocket.java:60) [prospero-2.37.0.jar:2.37.0]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor.lambda$sendTotal$2(DashboardUpdateProcessor.java:61) [prospero-2.37.0.jar:2.37.0]
       at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[na:1.8.0_231]
       at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1082) ~[na:1.8.0_231]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor.sendTotal(DashboardUpdateProcessor.java:61) [prospero-2.37.0.jar:2.37.0]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor.access$200(DashboardUpdateProcessor.java:11) [prospero-2.37.0.jar:2.37.0]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor$1.onDashboardTotalChange(DashboardUpdateProcessor.java:35) ~[prospero-2.37.0.jar:2.37.0]
       at com.myapp.storage.action.dashboard.DashboardTotal.sendUpdate(DashboardTotal.java:44) ~[prospero-2.37.0.jar:2.37.0]
       at com.myapp.storage.action.dashboard.Dashboard.sendUpdates(Dashboard.java:61) ~[prospero-2.37.0.jar:2.37.0]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateProcessor.processDashboard(DashboardUpdateProcessor.java:20) [prospero-2.37.0.jar:2.37.0]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateManager.lambda$checkForCellUpdates$0(DashboardUpdateManager.java:104) ~[prospero-2.37.0.jar:2.37.0]
       at java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1597) ~[na:1.8.0_231]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateManager.checkForCellUpdates(DashboardUpdateManager.java:94) ~[prospero-2.37.0.jar:2.37.0]
       at com.myapp.http.servlet.dashboard.updates.DashboardUpdateManager$UpdateDashboardsJob.execute(DashboardUpdateManager.java:154) ~[prospero-2.37.0.jar:2.37.0]
       at org.quartz.core.JobRunShell.run(JobRunShell.java:202) ~[quartz-2.3.1.jar:na]
       at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573) ~[quartz-2.3.1.jar:na]

```
Looking at the stacktrace, I can say that the first error is a reply to an inbound message (literally, the browser client sends "ping" and the server sends "pong").
The second error is our quartz-scheduled job sending updates to client.

This has happened only once (that I know of) in a year of operation.

Seems to be similar to #3279 and #2061 - however, I am unsure if its already supposed to be fixed in my version.

**Comment by superbazza** [comment]
Also, a follow-up question: Once this situation occurs, is it possible to recover from it by timing out the thread and starting again? Or are the websocket connections in a state where the only recovery is to re-start the JVM?

**Comment by sbordet** [comment]
@superbazza please update to 9.4.28, there are many things that may have been fixed between the version you are using and the latest one (your version is almost 2 years old now).

**Comment by lachlan-roberts** [comment]
This does look to be a duplicate of #2061, there is the same stack trace from `Deflater.ensureOpen()` triggering the issue. 

There are lots of other websocket bugs which have been fixed since 9.4.10 as well, so it is worth updating to the latest version of jetty.

**Comment by superbazza** [comment]
Excellent. Thanks for the response. Will upgrade now.
